cmd/k8s-operator/connector.go

```diff
@@ -139,55 +162,125 @@ func (a *ConnectorReconciler) Reconcile(ctx context.Context, req reconcile.Reque
 		}
 	}
 
-	// A Connector with unset .spec.subnetRouter and unset
-	// cn.spec.subnetRouter.Routes will be rejected at apply time (because
-	// these fields are set as required by our CRD validation). This check
-	// is here for if our CRD validation breaks unnoticed we don't crash the
-	// operator with nil pointer exception.
-	if cn.Spec.SubnetRouter == nil || len(cn.Spec.SubnetRouter.Routes) < 1 {
+	if err := a.validate(cn); err != nil {
+		logger.Errorf("error validating connector spec: %w", err)
 		return reconcile.Result{}, nil
 	}
 
-	if err := validateSubnetRouter(*cn.Spec.SubnetRouter); err != nil {
-		msg := fmt.Sprintf(messageSubnetRouterInvalid, err)
-		cn.Status.SubnetRouter = &tsapi.SubnetRouterStatus{
-			Ready:   metav1.ConditionFalse,
-			Reason:  reasonSubnetRouterInvalid,
-			Message: msg,
+	err = a.maybeProvisionComponent(ctx, logger, cn)
+	return reconcile.Result{}, err
+}
+
+// maybeProvisionSubnetRouter maybe deploys subnet router that exposes a subset of cluster cidrs to the tailnet
+func (a *ConnectorReconciler) maybeProvisionComponent(ctx context.Context, logger *zap.SugaredLogger, cn *tsapi.Connector) error {
+	typ, err := componentType(cn)
+	if err != nil {
+		msg := fmt.Sprintf("internal error determining component type: %v", err)
+		logger.Errorf(msg)
+		a.recorder.Eventf(cn, corev1.EventTypeWarning, reasonComponentCreationFailed, msg)
+		return nil
+	}
+	if typ == componentTypeExitNode {
+		cn.Status.IsExitNode = true
+	}
```

Comment:
These type of things in the `ConnectorReconciler` functionality are not elegant, but I was reluctant to define a higher level abstraction (i.e a `connector` interface) when there are still just two component types

Reply:
Out of date comment: has been removed with the refactor